src/main/java/mekanism/common/integration/framedblocks/ChemicalCamoContainerFactory.java

```diff
@@ -0,0 +1,143 @@
+
+final class ChemicalCamoContainerFactory extends CamoContainerFactory<ChemicalCamoContainer> {
+
+    private static final MapCodec<ChemicalCamoContainer> CODEC = Chemical.CODEC
+            .xmap(ChemicalCamoContainer::new, ChemicalCamoContainer::getChemical)
+            .fieldOf("chemical");
+    private static final StreamCodec<RegistryFriendlyByteBuf, ChemicalCamoContainer> STREAM_CODEC =
+            Chemical.STREAM_CODEC.map(ChemicalCamoContainer::new, ChemicalCamoContainer::getChemical);
+    private static final Component MSG_RADIOACTIVE = TextComponentUtil.translate(
+            MekanismLang.FRAMEDBLOCKS_CAMO_RADIOACTIVE.getTranslationKey()
+    );
+
+    @Override
+    protected void writeToNetwork(CompoundTag tag, ChemicalCamoContainer camo) {
+        Chemical chemical = camo.getChemical();
+        tag.putInt("chemical", MekanismAPI.CHEMICAL_REGISTRY.getId(chemical));
+    }
+
+    @Override
+    protected ChemicalCamoContainer readFromNetwork(CompoundTag tag) {
+        Chemical chemical = MekanismAPI.CHEMICAL_REGISTRY.byId(tag.getInt("chemical"));
+        return new ChemicalCamoContainer(chemical);
+    }
+
+    @Override
+    @Nullable
+    public ChemicalCamoContainer applyCamo(Level level, BlockPos pos, Player player, ItemStack stack) {
+        IChemicalHandler handler = Capabilities.CHEMICAL.getCapability(stack);
+        if (handler == null || handler.getChemicalTanks() <= 0) {
+             return null;
+        }
+
+        ChemicalStack chemical = handler.getChemicalInTank(0);
+        if (!isValidChemical(chemical.getChemical(), player)) {
+            return null;
+        }
+
+        if (!player.isCreative() && ConfigView.Server.INSTANCE.shouldConsumeCamoItem()) {
+            ChemicalStack extracted = handler.extractChemical(0, FramedBlocksIntegration.Constants.CHEMICAL_AMOUNT, Action.SIMULATE);
+            if (extracted.getAmount() != FramedBlocksIntegration.Constants.CHEMICAL_AMOUNT) {
+                return null;
+            }
+
+            if (!level.isClientSide()) {
+                handler.extractChemical(0, FramedBlocksIntegration.Constants.CHEMICAL_AMOUNT, Action.EXECUTE);
+            }
+        }
+
+        return new ChemicalCamoContainer(chemical.getChemical());
+    }
+
+    @Override
+    public boolean removeCamo(Level level, BlockPos pos, Player player, ItemStack stack, ChemicalCamoContainer camo) {
+        if (stack.isEmpty()) {
+            return false;
+        }
+
+        IChemicalHandler handler = Capabilities.CHEMICAL.getCapability(stack);
+        if (handler == null || handler.getChemicalTanks() <= 0) {
+            return false;
+        }
+
+        ChemicalStack chemical = camo.getChemical().getStack(FramedBlocksIntegration.Constants.CHEMICAL_AMOUNT);
+        if (handler.insertChemical(chemical, Action.SIMULATE).isEmpty()) {
+            if (!level.isClientSide() && !player.isCreative() && ConfigView.Server.INSTANCE.shouldConsumeCamoItem()) {
```

Comment:
Shouldn't we check if the camo item has to be consumed before trying to insert the chemical? Because if it doesn't need to be then we can theoretically just return true?

Reply:
I think you can do the simulate check then. And check if it fully fit for when it requires the resource, and just if any fit for when it doesn’t. Or maybe even just if it has that type stored? Not really sure.

Reply:
The latter is what I implemented. It has the benefit of allowing you to remove the camo with a full tank, which would fail if I tested it with a simulated 1mB insertion.